### PR TITLE
Import export stackup xml

### PR DESCRIPTION
**Pull request by ring630**

### CODE REVIEW

File: pyaedt/edb_core/stackup.py [hunk 1]
```diff
@@ -1271,6 +1226,421 @@ def residual_copper_area_per_layer(self):
...
+
+        return layers, materials, roughness_models
+
+    def _import_xml(self, file_path):
+        """Read external xml file and update stackup.
+
+        Parameters
+        ----------
+        file_path: str
+            Path to external xml file.
```
Comment (PipKat) [inline pyaedt/edb_core/stackup.py:1537]:
```suggestion
            Path to external XML file.
```

File: pyaedt/edb_core/stackup.py [hunk 2]
```diff
@@ -526,66 +528,7 @@ def remove_layer(self, name):
...
-            if name not in df.index:
-                self.remove_layer(name)
-        return True
-
-    @pyaedt_function_handler
-    def export_stackup(self, fpath, file_format="csv", include_material_with_layer=False):
+    def export_stackup(self, fpath, file_format="xml", include_material_with_layer=False):
         """Export stackup definition to a CSV or JSON file.
 
         Parameters
```
Comment (PipKat) [inline pyaedt/edb_core/stackup.py:534]: In line 537, use capital letters for file extension types (CVS or JSON, not csv or json). There are other occurrences that should also be changed. The rule is when you are specifying a specific file, surround in double back ticks to tag (such as ``test.csv``.  When the extension is used without a file name, you capitalize the extension.

File: pyaedt/edb_core/stackup.py [hunk 3]
```diff
@@ -1271,6 +1226,421 @@ def residual_copper_area_per_layer(self):
...
+                            self.stackup_layers[layer["name"]]._load_layer(layer)
+            return True
+
+    def _import_csv(self, file_path):
+        """Import stackup defnition from csv file.
+
+        Parameters
+        ----------
+        fpath : str
+            File path to csv or json file.
```
Comment (PipKat) [inline pyaedt/edb_core/stackup.py:1279]:
```suggestion
            File path to the CSV or JSON file.
```
Should JSON file be mentioned here? The method says from a CSV file.

File: pyaedt/edb_core/stackup.py [hunk 4]
```diff
@@ -1271,6 +1226,421 @@ def residual_copper_area_per_layer(self):
...
+        """
+        if not pd:
+            self._pedb.logger.error("Pandas is needed. Please, install it first.")
```
Comment (PipKat) [inline pyaedt/edb_core/stackup.py:1282]:
```suggestion
            self._pedb.logger.error("Pandas is needed. You must install it first.")
```

File: pyaedt/edb_core/stackup.py [hunk 5]
```diff
@@ -1271,6 +1226,421 @@ def residual_copper_area_per_layer(self):
...
+            return False
+        if os.path.splitext(file_path)[1] == ".json":
+            return self._import_layer_stackup(file_path)
+        if is_ironpython:
+            self._pedb.logger.error("Method working on CPython only.")
```
Comment (PipKat) [inline pyaedt/edb_core/stackup.py:1287]:
```suggestion
            self._pedb.logger.error("Method works on CPython only.")
```

File: pyaedt/edb_core/stackup.py [hunk 6]
```diff
@@ -1271,6 +1226,421 @@ def residual_copper_area_per_layer(self):
...
+                    layer.assign_roughness_model(
+                        "groisse", groisse_roughness=attr[attr_name]["Roughness"], apply_on_surface=on_surface
+                    )
+        return True
+
+    def _get(self):
+        """Get stackup information from layout.
+
+        Returns:
+        dict, dict, dict
```
Comment (PipKat) [inline pyaedt/edb_core/stackup.py:1471]: These dictionaries should be described individually.

File: pyaedt/edb_core/stackup.py [hunk 7]
```diff
@@ -1271,6 +1226,421 @@ def residual_copper_area_per_layer(self):
...
+            if layer_dict[name]["Type"] == "conductor":
+                layer_dict[name]["Type"] = "signal"
+
+            if list(l):
+                roughness_dict[name] = {i.tag: i.attrib for i in list(l)}
+
+        return self._set(layer_dict, material_dict, roughness_dict)
+
+    def _export_xml(self, file_path):
+        """Export stackup information to an external xml file.
```
Comment (PipKat) [inline pyaedt/edb_core/stackup.py:1573]:
```suggestion
        """Export stackup information to an external XMLfile.
```

File: pyaedt/edb_core/stackup.py [hunk 8]
```diff
@@ -1271,6 +1226,421 @@ def residual_copper_area_per_layer(self):
...
+
+        Parameters
+        ----------
+        file_path: str
+            Path to external xml file.
```
Comment (PipKat) [inline pyaedt/edb_core/stackup.py:1578]:
```suggestion
            Path to external XML file.
```

File: pyaedt/edb_core/stackup.py [hunk 9]
```diff
@@ -1271,6 +1226,421 @@ def residual_copper_area_per_layer(self):
...
+        Returns
+        -------
+        bool
+            ``True`` when succeed ``False`` if not.
```
Comment (PipKat) [inline pyaedt/edb_core/stackup.py:1582]:
```suggestion
            ``True`` when successful, ``False`` when failed.
```

File: pyaedt/edb_core/stackup.py [hunk 10]
```diff
@@ -1271,6 +1226,421 @@ def residual_copper_area_per_layer(self):
...
+            for pname, pval in val.items():
+                pval = {i: str(j) for i, j in pval.items()}
+                ET.SubElement(el, pname, pval)
+
+        write_pretty_xml(root, file_path)
+        return True
+
+    @pyaedt_function_handler
+    def import_stackup(self, file_path):
+        """Import stackup from file. File format can be xml, csv, json.
```
Comment (PipKat) [inline pyaedt/edb_core/stackup.py:1618]:
```suggestion
        """Import stackup from a file. The file format can be XML, CSV, or JSON.
```

File: pyaedt/edb_core/stackup.py [hunk 11]
```diff
@@ -1271,6 +1226,421 @@ def residual_copper_area_per_layer(self):
...
+
+        Parameters
+        ----------
+        file_path : str
+            File path to stackup file.
```
Comment (PipKat) [inline pyaedt/edb_core/stackup.py:1623]:
```suggestion
            Path to stackup file.
```

File: pyaedt/edb_core/stackup.py (hunk 6 above)
Comment (maxcapodi78) [reply pyaedt/edb_core/stackup.py]: Returns should be 
tuple of dict
 and then you describe the dicts

**Review by dcrawforAtAnsys — APPROVED** [review]
I tested and added some code to ignore non-physical layers which failed for my test case.

**Pull request merged**